### PR TITLE
Add a replace-type parameter to allow replacing package and/or type names

### LINKED ISSUE #331: Mockery follows type aliases and cannot compile mock
We have an interface with the following method
```go
import (
	"context"
	jet "github.com/go-jet/jet/mysql"
)

type Storage interface {
	BulkUpdateValues(ctx context.Context, primaryKeys []string, values map[jet.Column]interface{}) error
}
```

In the `jet/mysql` package, `Column` is defined as
```go
import "github.com/go-jet/jet/internal/jet"

// Column is common column interface for all types of columns.
type Column = jet.ColumnExpression
```

The generated mock uses `jet.ColumnExpression` from the `"github.com/go-jet/jet/internal/jet"` package, but since that package is internal to jet, it fails to compile.

```go
// Code generated by mockery v2.2.1. DO NOT EDIT.

package mocks

import (
	context "context"

	jet "github.com/go-jet/jet/internal/jet"

	mock "github.com/stretchr/testify/mock"
)

// Storage is an autogenerated mock type for the Storage type
type Storage struct {
	mock.Mock
}

// BulkUpdateValues provides a mock function with given fields: ctx, primaryKeys, values
func (_m *Storage) BulkUpdateValues(ctx context.Context, primaryKeys []string, values map[jet.ColumnExpression]interface{}) error {
	ret := _m.Called(ctx, primaryKeys, values)

	var r0 error
	if rf, ok := ret.Get(0).(func(context.Context, []string, map[jet.ColumnExpression]interface{}) error); ok {
		r0 = rf(ctx, primaryKeys, values)
	} else {
		r0 = ret.Error(0)
	}

	return r0
}
```

### PR DESCRIPTION
**Pull request by RangelReale**
Description
-------------
Add a `replace-type` parameter to allow replacing package and/or type names.

This overcomes the problem of type aliases to internal packages, like `cloud.google.com/go/pubsub.Message`, which is a type alias, for which the mock forwarded the implementation to an internal package that's not exported.

- Fixes #331 (issue)

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

Version of Golang used when building/testing:
---------------------------------------------

- [ ] 1.11
- [ ] 1.12
- [ ] 1.13
- [ ] 1.14
- [ ] 1.15
- [ ] 1.16
- [ ] 1.17
- [ ] 1.18
- [X] 1.19
- [ ] 1.20

How Has This Been Tested?
---------------------------

Tests included on the MR.
Tested with local code that imported `cloud.google.com/go/pubsub.Message`.

Checklist
-----------

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by LandonTClipp** [comment]
So I might be missing something here, but it does seem to me like the parsing packages Golang provides would indeed be able to differentiate type alias from the underlying type, as shown by my proof of concept. Although I wonder if the issue is that the `types.TypeName` passed to `getPackageScopedType` was resolved to the real type farther up the call stack.

The `types` package clearly understands the difference between alias and types: https://pkg.go.dev/go/types#TypeName

The question becomes, is the `golang.org/x/tools/go/packages` package not loading alias info? What led you to believe that the parser doesn't support alias information?

**Comment by RangelReale** [comment]
> So I might be missing something here, but it does seem to me like the parsing packages Golang provides would indeed be able to differentiate type alias from the underlying type, as shown by my proof of concept in the linked issue. Although I wonder if the issue is that the `types.TypeName` passed to `getPackageScopedType` was resolved to the real type farther up the call stack.
> 
> The `types` package clearly understands the difference between alias and types: https://pkg.go.dev/go/types#TypeName
> 
> The question becomes, is the `golang.org/x/tools/go/packages` package not loading alias info? What led you to believe that the parser doesn't support alias information?
> 
> If it _is_ possible to extract alias names, I would much prefer that solution than having the burden be placed on users to create manual exceptions to type names. I have a suspicion mockery is simply doing something wrong here.

I looked at the issues of other mock generator and they all have the same problem, see these 2 gomock issues for example:

[mockgen should not follow type aliases](https://github.com/golang/mock/issues/244)
[Document why a user should choose the source or reflect mode](https://github.com/golang/mock/issues/406)

Looks like they created an alternative "source" mode to parse the source directly to avoid missing this information. I stepped on the code and looked at the object returned to the type alias, I didn't find any mention of the alias anywhere, the same that is said on the first issue above.

So with the current mockery version it is effectivelly impossible to overcome the `cloud.google.com/go/pubsub.Message` problem, so this is an alternative, as this other implementation will be hard and time-consuming to do.

**Comment by LandonTClipp** [comment]
What I would like to determine before committing to this proposal is what part of the mockery call stack do we lose the alias information? I'm betting that the breakage is happening somewhere in this method: https://github.com/vektra/mockery/blob/8641a5bd64e0ea85e80b92dfd944259957092a3c/pkg/generator.go#L59

but I want to confirm _for certain_ that there's nothing we could do here. At the moment it's not clear to me at what point the alias information is lost.

**Comment by RangelReale** [comment]
> What I would like to determine before committing to this proposal is what part of the mockery call stack do we lose the alias information? I'm betting that the breakage is happening somewhere in this method:
> 
> https://github.com/vektra/mockery/blob/8641a5bd64e0ea85e80b92dfd944259957092a3c/pkg/generator.go#L59
> 
> but I want to confirm _for certain_ that there's nothing we could do here. At the moment it's not clear to me at what point the alias information is lost. I'm sure we could add a bunch of print statements in the code to prove it... it would help my sanity.

It is easy to step in the code using this MR to trigger this problem, step into the `generator_test.go:TestInternalPackagePrologue` function and you can see what happens to the `*types.TypeName` variable, I looked at it into a debugger and found no way to go back the the alias value itself.

**Comment by wyattanderson** [comment]
> So I might be missing something here, but it does seem to me like the parsing packages Golang provides would indeed be able to differentiate type alias from the underlying type, as shown by my proof of concept in the linked issue.

There's a gap in your proof-of-concept. You're correct in that you can identify a type alias _declaration_. However, aliases are not preserved in the type graph, so any variables, struct fields, etc. of an alias type are resolved as the _aliased type itself_ and any useful information about the alias itself is discarded and unavailable from those nodes. You can't distinguish between the alias and the type to which it refers when using `packages.Load`, etc., because the information simply isn't there. This, unfortunately, impacts numerous tools in the Go ecosystem beyond just Mockery.

This is the closest thing to an upstream tracking issue for this problem that I can find: https://github.com/golang/go/issues/44410

I'd love to have the proposed functionality as an option until the Go language itself fixes the underlying issue, and I'm not holding my breath waiting for that to happen.

**Comment by LandonTClipp** [comment]
@wyattanderson that makes more sense. I understand the need for this a little better. It's unfortunate and I really hate having to maintain hacky stuff like this but I can't think of a better solution until the go maintainers fix the parsers (which probably isn't anywhere high on their priority list).

After having given it more thought I think we can go ahead with this. However @RangelReale could you add a section in the mkdocs page here: https://vektra.github.io/mockery/features/ . We can just add another section regarding type aliases and how to resolve it with the feature you've implemented here.

**Comment by RangelReale** [comment]
@LandonTClipp yes I fully agree with you, I hate having to add these kinds of workarounds, but in this case there seems to be absolutely no way around it. I'll look at cleaning up the code and updating the documentation tomorrow.

@wyattanderson @LandonTClipp maybe we can detect that this is an alias and issue a warning about it?

**Comment by RangelReale** [comment]
I refactored the parameter parsing to be done only at startup, see if everything is ok.

**Comment by LandonTClipp** [comment]
Great work, I’ll give it a review in the next day or so.

**Review by LandonTClipp — CHANGES_REQUESTED** [review]
I have some changes I want to make, one of which will hopefully simplify the unit test a bit. Let me know what you think of those.

File: pkg/generator_test.go [hunk 1]
```diff
@@ -2411,6 +2411,93 @@ import mock "github.com/stretchr/testify/mock"
...
+	s.checkPrologueGeneration(generator, expected)
+}
+
+func (s *GeneratorSuite) TestInternalPackage() {
+	expected := `// Foo is an autogenerated mock type for the Foo type
+type Foo struct {
+	mock.Mock
+}
+
+// DoFoo provides a mock function with given fields:
```
[the remark this answers is not in the document]
Comment (RangelReale) [reply pkg/generator_test.go]: I added a regexp check function and used it for this assertion.

**Comment by LandonTClipp** [comment]
Thank you sir, I'll take another look at this in the next day or so!

**Review by LandonTClipp — APPROVED** [review]

**Comment by papadeltasierra** [comment]
Very late to the game but hit this error and I have a problem that this does not solve.  I have multiple packages that `mockery` maps down to a single internal package and I cannot use this parameter to replace back to those packages.  So what happens is:
- `mockery` follows packages `A`, `B`, `C` to internal package `Z`
- I can use `replace-type` to map `Z` to `A`, or `B`, or `C` but not all of them!
The package that causes this is not mine and I cannot fix it so seem unable to use `mockery` to mock my interface.

**Comment by LandonTClipp** [comment]
@papadeltasierra thats really tricky. Unfortunately the only way you can get around this is by manually specifying the replacements for each individual type. There is no way mockery can know where your type alias originally came from because historically, a type alias is not represented in the type tree. 

There is a lead I'm chasing up in Go 1.22 where we might be able to retain the alias information if you set a specific GODEBUG flag. The Go authors know this is an issue and have made some half attempt at providing a fix.

**Comment by LandonTClipp** [comment]
@papadeltasierra yes the high level explanation is that prior to Go 1.22, the parsed type tree did not have any representation for a type alias. It always got resolved down to the underlying type, so there was no way to extract the alias information. This was a limitation with the Go package parser itself, not with mockery.

You can see in 1.22 they added a [`types.Alias`](https://pkg.go.dev/go/types@go1.22.0#Alias) type that would get returned when gotypelias is set to 1. This is probably the solution moving forward, but we need to add support for this in mockery.

**Comment by LandonTClipp** [comment]
@papadeltasierra update on the `gotypelias` lead, we will not be able to get support for handling this cleanly until at least Go 1.23 when the Go devs make the requisite changes to `golang.org/x/tools/go/packages`. See more details here: https://github.com/vektra/mockery/issues/763#issuecomment-1955431577

**Pull request merged**